File: tools/convert_wav_words_to_m4a.swift
#!/usr/bin/env swift

import AVFoundation
import Foundation

struct Arguments {
    let audioDir: URL
    let overwrite: Bool
    let keepWav: Bool
    let dryRun: Bool
}

enum ScriptError: Error, CustomStringConvertible {
    case invalidArguments(String)
    case conversionFailed(String)
    case exportFailed(String)
    case manifestMissing

    var description: String {
        switch self {
        case .invalidArguments(let message): return message
        case .conversionFailed(let message): return message
        case .exportFailed(let message): return message
        case .manifestMissing: return "manifest.json not found"
        }
    }
}

func parseArguments() throws -> Arguments {
    let args = Array(CommandLine.arguments.dropFirst())
    var audioDir = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
        .appendingPathComponent("shared/src/commonMain/composeResources/files/audio")
    var overwrite = false
    var keepWav = true
    var dryRun = false

    var index = 0
    while index < args.count {
        let arg = args[index]
        switch arg {
        case "--audio-dir":
            index += 1
            guard index < args.count else {
                throw ScriptError.invalidArguments("--audio-dir requires a value")
            }
            audioDir = URL(fileURLWithPath: args[index], relativeTo: URL(fileURLWithPath: FileManager.default.currentDirectoryPath)).standardizedFileURL
        case "--overwrite":
            overwrite = true
        case "--delete-wav":
            keepWav = false
        case "--dry-run":
            dryRun = true
        case "--help":
            print("Usage: swift tools/convert_wav_words_to_m4a.swift [--audio-dir <path>] [--overwrite] [--delete-wav] [--dry-run]")
            exit(0)
        default:
            throw ScriptError.invalidArguments("Unknown argument: \(arg)")
        }
        index += 1
    }

    return Arguments(audioDir: audioDir, overwrite: overwrite, keepWav: keepWav, dryRun: dryRun)
}

func listWavFiles(in wordsDir: URL) throws -> [URL] {
    let fileManager = FileManager.default
    let entries = try fileManager.contentsOfDirectory(at: wordsDir, includingPropertiesForKeys: nil)
    return entries
        .filter { $0.pathExtension.lowercased() == "wav" }
        .sorted { $0.lastPathComponent < $1.lastPathComponent }
}

func convertWavToM4a(inputURL: URL, outputURL: URL) throws {
    let asset = AVURLAsset(url: inputURL)
    guard let audioTrack = asset.tracks(withMediaType: .audio).first else {
        throw ScriptError.conversionFailed("No audio track found in \(inputURL.lastPathComponent)")
    }

    let reader = try AVAssetReader(asset: asset)
    let readerOutput = AVAssetReaderTrackOutput(track: audioTrack, outputSettings: nil)
    guard reader.canAdd(readerOutput) else {
        throw ScriptError.conversionFailed("Cannot add reader output for \(inputURL.lastPathComponent)")
    }
    reader.add(readerOutput)

    try? FileManager.default.removeItem(at: outputURL)
    let writer = try AVAssetWriter(outputURL: outputURL, fileType: .m4a)
    guard let outputSettings = audioTrack.recommendedAudioSettingsForAssetWriter(writingTo: .m4a) as? [String: Any] else {
        throw ScriptError.conversionFailed("Cannot create AAC output settings for \(inputURL.lastPathComponent)")
    }
    let writerInput = AVAssetWriterInput(
        mediaType: .audio,
        outputSettings: outputSettings
    )
    writerInput.expectsMediaDataInRealTime = false
    guard writer.canAdd(writerInput) else {
        throw ScriptError.conversionFailed("Cannot add writer input for \(inputURL.lastPathComponent)")
    }
    writer.add(writerInput)

    guard reader.startReading() else {
        throw ScriptError.conversionFailed("Reader failed to start for \(inputURL.lastPathComponent)")
    }
    guard writer.startWriting() else {
        throw ScriptError.conversionFailed("Writer failed to start for \(inputURL.lastPathComponent)")
    }

    writer.startSession(atSourceTime: .zero)

    let semaphore = DispatchSemaphore(value: 0)
    let queue = DispatchQueue(label: "wav-to-m4a-writer")

    writerInput.requestMediaDataWhenReady(on: queue) {
        while writerInput.isReadyForMoreMediaData {
            if let sampleBuffer = readerOutput.copyNextSampleBuffer() {
                if !writerInput.append(sampleBuffer) {
                    writerInput.markAsFinished()
                    reader.cancelReading()
                    semaphore.signal()
                    return
                }
            } else {
                writerInput.markAsFinished()
                writer.finishWriting {
                    semaphore.signal()
                }
                return
            }
        }
    }

    semaphore.wait()

    if reader.status == .failed {
        throw ScriptError.conversionFailed("\(inputURL.lastPathComponent): \(reader.error?.localizedDescription ?? "Reader failed")")
    }
    if writer.status == .failed {
        throw ScriptError.exportFailed("\(inputURL.lastPathComponent): \(writer.error?.localizedDescription ?? "Writer failed")")
    }
}

func updateManifest(at manifestURL: URL) throws {
    let data = try Data(contentsOf: manifestURL)
    guard var manifest = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
        throw ScriptError.manifestMissing
    }

    manifest["generatedAt"] = ISO8601DateFormatter().string(from: Date())
    manifest["format"] = [
        "container": "m4a",
        "codec": "aac-lc",
        "sampleRateHz": 24000,
        "channels": 1,
        "bitrateKbps": 48
    ]

    if var items = manifest["items"] as? [[String: Any]] {
        items = items.map { item in
            var next = item
            if let wordAudio = item["wordAudio"] as? String {
                next["wordAudio"] = wordAudio.replacingOccurrences(of: ".wav", with: ".m4a")
            }
            if let examples = item["examples"] as? [[String: Any]] {
                next["examples"] = examples.map { example in
                    var nextExample = example
                    if let audio = example["audio"] as? String {
                        nextExample["audio"] = audio.replacingOccurrences(of: ".wav", with: ".m4a")
                    }
                    return nextExample
                }
            }
            return next
        }
        manifest["items"] = items
    }

    let outputData = try JSONSerialization.data(withJSONObject: manifest, options: [.prettyPrinted, .sortedKeys])
    try outputData.write(to: manifestURL)
    if let handle = try? FileHandle(forWritingTo: manifestURL) {
        try handle.seekToEnd()
        try handle.write(contentsOf: Data([0x0A]))
        try handle.close()
    }
}

do {
    let arguments = try parseArguments()
    let wordsDir = arguments.audioDir.appendingPathComponent("words")
    let manifestURL = arguments.audioDir.appendingPathComponent("manifest.json")

    guard FileManager.default.fileExists(atPath: manifestURL.path) else {
        throw ScriptError.manifestMissing
    }

    let wavFiles = try listWavFiles(in: wordsDir)
    if arguments.dryRun {
        let sample = wavFiles.prefix(5).map(\.lastPathComponent)
        let payload: [String: Any] = [
            "audioDir": arguments.audioDir.path,
            "wavFiles": wavFiles.count,
            "sample": sample
        ]
        let data = try JSONSerialization.data(withJSONObject: payload, options: [.prettyPrinted, .sortedKeys])
        print(String(decoding: data, as: UTF8.self))
        exit(0)
    }

    var converted = 0
    for wavURL in wavFiles {
        let m4aURL = wavURL.deletingPathExtension().appendingPathExtension("m4a")
        if FileManager.default.fileExists(atPath: m4aURL.path), !arguments.overwrite {
            continue
        }

        try convertWavToM4a(inputURL: wavURL, outputURL: m4aURL)
        if !arguments.keepWav {
            try? FileManager.default.removeItem(at: wavURL)
        }
        converted += 1
        if converted % 50 == 0 {
            print("Converted \(converted)/\(wavFiles.count)")
        }
    }

    try updateManifest(at: manifestURL)
    print("Converted \(converted) files")
    print("Updated \(manifestURL.path)")
} catch {
    fputs("\(error)\n", stderr)
    exit(1)
}

File: tools/convert_wav_words_to_m4a.mjs
#!/usr/bin/env node

import { mkdir, readFile, readdir, rm, stat, writeFile } from "node:fs/promises";
import path from "node:path";
import { spawn } from "node:child_process";

const DEFAULT_AUDIO_DIR = "shared/src/commonMain/composeResources/files/audio";
const DEFAULT_BITRATE_KBPS = 48;

function parseArgs(argv) {
    const args = {
        audioDir: DEFAULT_AUDIO_DIR,
        overwrite: false,
        keepWav: true,
        dryRun: false
    };

    for (let index = 0; index < argv.length; index += 1) {
        const arg = argv[index];
        const next = argv[index + 1];

        switch (arg) {
            case "--audio-dir":
                args.audioDir = next;
                index += 1;
                break;
            case "--overwrite":
                args.overwrite = true;
                break;
            case "--delete-wav":
                args.keepWav = false;
                break;
            case "--dry-run":
                args.dryRun = true;
                break;
            case "--help":
                printHelp();
                process.exit(0);
            default:
                throw new Error(`Unknown argument: ${arg}`);
        }
    }

    return args;
}

function printHelp() {
    console.log(`Usage:
  node tools/convert_wav_words_to_m4a.mjs [--audio-dir <path>] [--overwrite] [--delete-wav] [--dry-run]
`);
}

async function convertWavToM4a(inputPath, outputPath) {
    await new Promise((resolve, reject) => {
        const child = spawn("/usr/bin/afconvert", [
            "-f", "m4af",
            "-d", "aac",
            "-u", "vbrq", "27",
            "-u", "src", "c=1,r=24000",
            inputPath,
            outputPath
        ], {
            stdio: ["ignore", "pipe", "pipe"]
        });

        let stderr = "";
        child.stderr.on("data", (chunk) => {
            stderr += chunk.toString();
        });
        child.on("error", reject);
        child.on("close", (code) => {
            if (code === 0) {
                resolve();
                return;
            }
            reject(new Error(`afconvert failed for ${path.basename(inputPath)}: ${stderr.trim()}`));
        });
    });
}

async function listWordWavFiles(wordsDir) {
    const entries = await readdir(wordsDir, { withFileTypes: true });
    return entries
        .filter((entry) => entry.isFile() && entry.name.endsWith(".wav"))
        .map((entry) => path.join(wordsDir, entry.name))
        .sort();
}

async function updateManifest(manifestPath) {
    const manifest = JSON.parse(await readFile(manifestPath, "utf8"));
    manifest.generatedAt = new Date().toISOString();
    manifest.format = {
        container: "m4a",
        codec: "aac-lc",
        sampleRateHz: 24000,
        channels: 1,
        bitrateKbps: DEFAULT_BITRATE_KBPS
    };

    if (Array.isArray(manifest.items)) {
        manifest.items = manifest.items.map((item) => ({
            ...item,
            wordAudio: typeof item.wordAudio === "string"
                ? item.wordAudio.replace(/\.wav$/i, ".m4a")
                : item.wordAudio,
            examples: Array.isArray(item.examples)
                ? item.examples.map((example) => ({
                    ...example,
                    audio: typeof example.audio === "string"
                        ? example.audio.replace(/\.wav$/i, ".m4a")
                        : example.audio
                }))
                : []
        }));
    }

    await writeFile(`${manifestPath}`, `${JSON.stringify(manifest, null, 2)}\n`, "utf8");
}

async function main() {
    const args = parseArgs(process.argv.slice(2));
    const audioDir = path.resolve(process.cwd(), args.audioDir);
    const wordsDir = path.join(audioDir, "words");
    const manifestPath = path.join(audioDir, "manifest.json");

    await mkdir(wordsDir, { recursive: true });
    await stat(manifestPath);

    const wavFiles = await listWordWavFiles(wordsDir);
    if (args.dryRun) {
        console.log(JSON.stringify({
            audioDir,
            wavFiles: wavFiles.length,
            sample: wavFiles.slice(0, 5).map((file) => path.basename(file))
        }, null, 2));
        return;
    }

    let converted = 0;
    for (const wavPath of wavFiles) {
        const m4aPath = wavPath.replace(/\.wav$/i, ".m4a");
        try {
            if (!args.overwrite) {
                await stat(m4aPath);
                continue;
            }
        } catch {
            // output does not exist
        }

        await convertWavToM4a(wavPath, m4aPath);
        if (!args.keepWav) {
            await rm(wavPath, { force: true });
        }
        converted += 1;
        if (converted % 50 === 0) {
            console.log(`Converted ${converted}/${wavFiles.length}`);
        }
    }

    await updateManifest(manifestPath);
    console.log(`Converted ${converted} files`);
    console.log(`Updated ${manifestPath}`);
}

main().catch((error) => {
    console.error(error.message);
    process.exit(1);
});
